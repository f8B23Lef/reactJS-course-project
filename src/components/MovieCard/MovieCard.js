import React from 'react';
import PropTypes from 'prop-types';
import ThreeDotsMenu from '../ThreeDotsMenu/ThreeDotsMenu';
import { getYearFromDate } from '../../utils/utils';
import { useHover } from '../../hooks/index';
import './MovieCard.scss';

export default function MovieCard(props) {
  const {
    movie,
    onEdit,
    onDelete,
    onClick,
  } = props;

  const [hoverRef, isHovered] = useHover();

  const tdmItems = [{
    itemText: 'Edit',
    onClick: onEdit.bind(null, movie.id),
  }, {
    itemText: 'Delete',
    onClick: onDelete.bind(null, movie.id),
  }];

  return (
    <div
      className='movie-card'
      data-testid='movie-card'
      ref={hoverRef}
      onClick={() => onClick(movie.id)}
    >
      <img src={movie.poster_path} alt='movie poster' />
      <ThreeDotsMenu
        isHovered={isHovered}
        tdmItems={tdmItems}
      />
      <div className='movie-summary'>
        <div>
          <p className='title'>{movie.title}</p>
          <p className='description'>{movie.genres.join(', ')}</p>
        </div>
        <div className='year'>{getYearFromDate(movie.release_date)}</div>
      </div>
    </div>
  );
}

MovieCard.propTypes = {
  movie: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string,
    release_date: PropTypes.string,
    vote_average: PropTypes.number,
    genres: PropTypes.arrayOf(PropTypes.string),
    runtime: PropTypes.number,
    overview: PropTypes.string,
    poster_path: PropTypes.string.isRequired,
  }),
  onEdit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onClick: PropTypes.func.isRequired,
};

MovieCard.defaultProps = {
  movie: {
    title: '',
    release_date: '',
    poster_path: '',
    vote_average: 0,
    genres: [],
    runtime: 0,
    overview: '',
  },
};
