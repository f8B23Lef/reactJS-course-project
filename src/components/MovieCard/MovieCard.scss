@import '../../styles/index.scss';

.movie-card {
  flex: 0 0 200px;
  margin: 0 35px 20px 0;
  position: relative;

  img {
    width: 100%;
    height: 300px;
  }

  .icon {
    display: none;
    position: absolute;
    top: 5px;
    right: 8px;
    width: 22px;
    height: 22px;
    border: 1px solid $dark-grey;
    border-radius: 50%;
    background-color: $dark-grey;
  }

  &:hover {
    .icon {
      display: block;
    }
  }
}

.movie-summary {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  padding: 5px 0;
  color: $white;
  opacity: 0.8;

  .title {
    margin: 0;
    padding: 0 0 5px 0;
    font-size: 14px;
  }

  .description {
    margin: 0;
    font-size: 12px;
    opacity: 0.6;
  }
  
  .year {
    padding: 2px 5px;
    height: fit-content;
    font-size: 12px;
    border: 1px solid;
    border-radius: 2px;
  }
}
