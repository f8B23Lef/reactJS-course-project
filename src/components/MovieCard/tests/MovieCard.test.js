import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { render, screen, fireEvent } from '@testing-library/react';
import MovieCard from '../MovieCard';

jest.mock(
  '../../../components/ThreeDotsMenu/ThreeDotsMenu',
  () => () => 'ThreeDotsMenuComponent',
);

describe('MovieCard', () => {
  let movie;

  beforeEach(() => {
    movie = {
      id: 1,
      title: 'Movie1',
      genres: ['Crime', 'Action'],
      release_date: '2022-10-01',
      poster_path: '/poster',
    };
  });

  it('renders MovieCard snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <MovieCard
          movie={movie}
          onEdit={jest.fn()}
          onDelete={jest.fn()}
          onClick={jest.fn()}
        />
      </BrowserRouter>,
    );

    expect(asFragment(
      <MovieCard
        movie={movie}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
        onClick={jest.fn()}
      />,
    )).toMatchSnapshot();
  });

  it('renders image with specific path', () => {
    render(
      <MovieCard
        movie={movie}
        onEdit={jest.fn()}
        onDelete={jest.fn()}
        onClick={jest.fn()}
      />,
    );
    const imageElement = screen.getByAltText('movie poster');

    expect(imageElement).toHaveAttribute('src', '/poster');
  });

  it('calls onClick() prop with movie id param on movie card click', () => {
    const onClickMock = jest.fn();
    render(
      <BrowserRouter>
        <MovieCard
          movie={movie}
          onEdit={jest.fn()}
          onDelete={jest.fn()}
          onClick={onClickMock}
        />
      </BrowserRouter>,
    );

    const cardElement = screen.getByTestId('movie-card');
    fireEvent.click(cardElement);

    expect(onClickMock).toHaveBeenCalledWith(movie.id);
  });
});
