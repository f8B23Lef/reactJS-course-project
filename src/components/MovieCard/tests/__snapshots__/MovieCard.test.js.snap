// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard renders MovieCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="movie-card"
    data-testid="movie-card"
  >
    <img
      alt="movie poster"
      src="/poster"
    />
    ThreeDotsMenuComponent
    <div
      class="movie-summary"
    >
      <div>
        <p
          class="title"
        >
          Movie1
        </p>
        <p
          class="description"
        >
          Crime, Action
        </p>
      </div>
      <div
        class="year"
      >
        2022
      </div>
    </div>
  </div>
</DocumentFragment>
`;
